brooklyn-server/core/src/main/java/org/apache/brooklyn/core/entity/AbstractApplication.java

```diff
@@ -220,13 +221,37 @@ public void stop() {
                 //but that could be handled by the impl at management
                 //(keeping recently unmanaged things)  
                 //  however unmanaging must be done last, _after_ we stop children and set attributes 
+                Location machineProvisioningLocation = Iterables.getFirst(getLocations(), null);
                 getEntityManager().unmanage(this);
+                if (machineProvisioningLocation != null) {
+                    unmanageProvisionerIfNotUsed(machineProvisioningLocation);
+                }
             }
         }
 
         logApplicationLifecycle("Stopped");
     }
 
+    //TODO Think of general approach for expunging location provisioners.
+    // I left a code which could be a possible general approach. However it needs tests.
+    protected boolean unmanageProvisionerIfNotUsed(/*MachineProvisioningLocation<MachineLocation>*/Location machineProvisioningLocation) {
+        if (machineProvisioningLocation.getChildren().isEmpty()) {
+            Entity parentEntity = getParent();
+            if (parentEntity == null) {
+                getManagementContext().getLocationManager().unmanage(machineProvisioningLocation);
```

Comment:
This doesn't feel right - it wasn't the application that created the location, so should it be the application that unmanages the location? If we do this, it would be good to have some of indicating that the location is "owned" by the app, rather than it just having been passed to the new application when calling `start(location)`.